## **Maestría en Inteligencia Artificial Aplicada**
### **Curso: Inteligencia Artificial y Aprendizaje Automático**
#### Tecnológico de Monterrey
####P rof Luis Eduardo Falcón Morales

### **Adtividad de la Semana 8**
#### **Reducción de Dimensionalidad - Descomposición en Valores Singulares (SVD) y Sistemas de Recomendación**


**Esta Actividad deberás resolverla de manera individual.**

**Nombre y matrícula:**

* Juan Manuel Carballo
* A01166758


In [1]:
# Incluye aquí todas las librerías y paquetes que requieras.

import pandas as pd
import numpy as np


from sklearn.decomposition import TruncatedSVD


### **Liga de Datos de la UCI "Restaurant & consumer data".**
### **Descarga "RCdata.zip" y de ahí necesitaremos los archivos "rating_final.csv" y "geoplaces2.csv".**

https://archive.ics.uci.edu/dataset/232/restaurant+consumer+data



In [2]:
# Descarga los archivos de la página de la UCI a partir de los cuales generaremos
# nuestras matrices de utilidad.

data1 = pd.read_csv("../datasets/rating_final.csv", header='infer', sep=",")
data2 = pd.read_csv("../datasets/geoplaces2.csv", header='infer',  encoding='latin-1')

print(data1.shape, data2.shape)


(1161, 5) (130, 21)


## **Ejercicio - 1**

### **Explica cuál es el propósito de usar el argumento "latin-1" al cargar el segundo archivo.**

++++++++ Inicia la sección de agregar texto: +++++++++++

Los archivos al usar una codificacion como latin-1 los caracteres como acentos o letras con tildes se encuentran en una codificacion diferente y la mayoria de los lenguajes de alto nivel como python espera que el texto plano se encuentre codificado en UTF-8, por eso hay que especificar que tipo de codificacion se encuentra si el texto en el archivo es diferente del que normalmente acepta python.


++++++++ Termina la sección de agregar texto. +++++++++++

In [3]:
# Del primer archivo obtenemos una matriz con 3 evaluaciones de los restaurantes:
# general, comida y servicio.
# Las evaluaciones pueden ser 0, 1 o 2, siendo 0 la menor calificación y 2 la
# mayor calificación:

data1.head()

userID  placeID  rating  food_rating  service_rating
0  U1077   135085       2            2               2
1  U1077   135038       2            2               1
2  U1077   132825       2            2               2
3  U1077   135060       1            2               2
4  U1068   135104       1            1               2

In [4]:
# Del segundo archivo obtenemos información diversa de cada restaurante:

data2.head(2)

placeID   latitude   longitude  \
0   134999  18.915421  -99.184871   
1   132825  22.147392 -100.983092   

                                      the_geom_meter             name  \
0  0101000020957F000088568DE356715AC138C0A525FC46...  Kiku Cuernavaca   
1  0101000020957F00001AD016568C4858C1243261274BA5...  puesto de tacos   

                                  address        city    state country fax  \
0                              Revolucion  Cuernavaca  Morelos  Mexico   ?   
1  esquina santos degollado y leon guzman      s.l.p.   s.l.p.  mexico   ?   

   ...            alcohol smoking_area dress_code     accessibility   price  \
0  ...  No_Alcohol_Served         none   informal  no_accessibility  medium   
1  ...  No_Alcohol_Served         none   informal        completely     low   

                     url Rambience franchise    area other_services  
0  kikucuernavaca.com.mx  familiar         f  closed           none  
1                      ?  familiar         f    open           none  

[2 rows x 21 columns]

In [5]:
# De data1 no requeriremos "rating" y de data2 solo necesitamos "placeID" y "name":
# Definimos la lista y matrices con los factores que necesitamos:

lista_data1 = ['userID','placeID','food_rating','service_rating']
lista_data2 = ['placeID','name']

data1a = data1[lista_data1]
data2a = data2[lista_data2]

## **Ejercicio - 2**

In [7]:
# Define el DataFrame que conjunte la información de las dos DataFrame data1a
# y data2a en uno solo, a través de la columna común "placeID" y que
# llamaremos "df_combinado":



# ************* Inlcuye aquí tu código:*****************************


df_combinado = pd.merge(data1a, data2a, on='placeID')


# *********** Aquí termina la sección de agregar código *************


# Despleguemos la dimensión y los primeros renglones de este DataFrame:

print(df_combinado.shape)
df_combinado.head()

(1161, 5)


userID  placeID  food_rating  service_rating                    name
0  U1077   135085            2               2  Tortas Locas Hipocampo
1  U1108   135085            2               1  Tortas Locas Hipocampo
2  U1081   135085            2               1  Tortas Locas Hipocampo
3  U1056   135085            2               2  Tortas Locas Hipocampo
4  U1134   135085            1               2  Tortas Locas Hipocampo

## **Ejercicio - 3**

In [8]:
#    Define la matriz de utilidad cuyos renglones sean los nombres de los
#    restaurantes, las columnas los IDs de los usuarios y las entradas la
#    evaluación de la comida (food_rating). La llamaremos "UtMx_food".


# ************* Inlcuye aquí tu código:*****************************


UtMx_food = df_combinado.pivot_table(values='food_rating', index="name", columns="userID", fill_value=0)


# *********** Aquí termina la sección de agregar código *************


print('Dimensión de la matriz de Utilidad sobre la evaluación de la comida:')
print('(restaurantes, usuarios) =', (UtMx_food.shape))
UtMx_food.head()

Dimensión de la matriz de Utilidad sobre la evaluación de la comida:
(restaurantes, usuarios) = (129, 138)


userID                     U1001  U1002  U1003  U1004  U1005  U1006  U1007  \
name                                                                         
Abondance Restaurante Bar    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
Arrachela Grill              0.0    0.0    0.0    0.0    0.0    0.0    0.0   
Cabana Huasteca              0.0    0.0    2.0    0.0    0.0    0.0    0.0   
Cafe Chaires                 0.0    0.0    0.0    0.0    0.0    0.0    1.0   
Cafeteria cenidet            0.0    0.0    0.0    0.0    0.0    0.0    0.0   

userID                     U1008  U1009  U1010  ...  U1129  U1130  U1131  \
name                                            ...                        
Abondance Restaurante Bar    1.0    0.0    0.0  ...    0.0    0.0    0.0   
Arrachela Grill              0.0    0.0    0.0  ...    0.0    0.0    0.0   
Cabana Huasteca              0.0    2.0    0.0  ...    0.0    0.0    0.0   
Cafe Chaires                 0.0    0.0    0.0  ...    0.0    0.0    0.0   
Cafeteria cenidet            0.0    0.0    0.0  ...    0.0    0.0    0.0   

userID                     U1132  U1133  U1134  U1135  U1136  U1137  U1138  
name                                                                        
Abondance Restaurante Bar    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
Arrachela Grill              0.0    0.0    0.0    0.0    0.0    0.0    0.0  
Cabana Huasteca              0.0    0.0    1.0    0.0    0.0    0.0    0.0  
Cafe Chaires                 0.0    0.0    0.0    0.0    0.0    0.0    0.0  
Cafeteria cenidet            0.0    0.0    0.0    0.0    0.0    0.0    0.0  

[5 rows x 138 columns]

## **Ejercicio - 4**

### **En la Factorización SVD la cantidad de valores singulares será menor o igual al menor valor de los renglones o columnas de la matriz de utilidad UtMx.**

### **La factorización SVD de una matriz $A$ tiene la forma:**

$A_{m\times n} = U_{m\times m}\Sigma_{m\times n}V_{n\times n}^T$

In [12]:
# Apliquemos la transformación SVD para obtener la matriz de las variables
# latentes de los restaurantes en relación a la evalución de los usuarios con
# respecto a la comida. Deberás obtener la factorización SVD con el máximo de
# comonentes y por lo tanto el máximo de valores singulares posible.
# Asigna a la variable "nc_food" el valor del máximo número de valores singulares
# que podemos obtener de la utilidad que generamos en el ejercicio anterior.


# ************* Inlcuye aquí tu código:*****************************


nc_food = 138     # número de componentes


# *********** Aquí termina la sección de agregar código *************


# Inicializamos y aplicamos la factorización SVD en relación a la
# evaluación de la comida, para determinar la cantidad de componentes
# que podemos utilizar con un 90% de explicación de la variabilidad
# de dichas componentes:

SVD_food = TruncatedSVD( n_components = nc_food )
SVD_food.fit(UtMx_food)

# Determinemos la cantidad de componentes que explican un 90%
# de la variabilidad acumulada de cada componente:

for j in range(nc_food):
  if SVD_food.explained_variance_ratio_[0:j].sum() > 0.90:
    break

# Usaremos esta cantidad de componentes para las recomendaciones basadas
# en la calificación de la comida:
N_food = j-1

print('Total de valores singulares basados en evaluación de la :', nc_food)
print('Valor de truncamiento al 90% de dicha variabilidad:',N_food)


Total de valores singulares basados en evaluación de la : 138
Valor de truncamiento al 90% de dicha variabilidad: 51


In [13]:
# Usando la métrica de correlación de Pearson, obtendremos las 10 mejores
# recomendaciones de este modelo no supervisado con base a la información de
# alguien que desea obtener las similidades com el restaurante "Restaurante Pueblo
# Bonito" y con la cota de truncamiento obtenido en el ejercicio anterior.
# Para ello obtendremos las correlaciones positivas y de ahí deberás seleccionar
# las diez mejores.

# Factorización SVD:
SVD_food = TruncatedSVD(n_components = N_food)
resultant_matrix_food = SVD_food.fit_transform(UtMx_food)

# Matriz de correlación de Pearson:
corr_mat_food = np.corrcoef(resultant_matrix_food)

# Restaurante de referencia:
restaurante_de_referencia = "Restaurante Pueblo Bonito"
nombres_rest = UtMx_food.T.columns  # nombres de restaurantes
idx_rest = list(nombres_rest).index(restaurante_de_referencia)
corr_rest = corr_mat_food[idx_rest] # Vector de Correlación del RinconHuasteco

# Buscando las correlaciones positivas:
idx = (corr_rest>0)
mejores_sim_food = list()
for i in range(len(nombres_rest[idx])):
  mejores_sim_food.append((corr_rest[idx][i], nombres_rest[idx][i]))

print('Total de similaridades positivas encontradas:', len(mejores_sim_food))

print('Algunos de los resultados encontrados:')
mejores_sim_food[0:7]

Total de similaridades positivas encontradas: 72
Algunos de los resultados encontrados:


[(0.0030585018402347402, 'Arrachela Grill'),
 (0.022492414952969567, 'Cabana Huasteca'),
 (0.030893755479017713, 'Cafe Chaires'),
 (0.01693901426711723, 'Cafeteria cenidet'),
 (0.22446161595647848, 'Cafeteria y Restaurant El Pacifico'),
 (0.014924666678263398, 'Cenaduria El Rincón de Tlaquepaque'),
 (0.0007525853282054746, 'Chilis Cuernavaca')]

## **Ejercicio - 5**

In [32]:
# Ordenar la lista de recomendaciones "mejores_sim_food" encontrada en el paso
# anterior de manera descendente y llamaremos "mejores_sim_food_ordenadas".
#
# NOTA: Puedes incluir a continuación las líneas de código que requieras.

# ************* Inlcuye aquí tu código:*****************************



mejores_sim_food_ordenadas = sorted(mejores_sim_food, reverse=True)



# *********** Aquí termina la sección de agregar código *************


# Desplegamos las 10 mejores similitudes encontradas de manera descendente:
print('Similitudes con base a la evaluación de la comida con mayores valores de correlación:')
for k in range(1,11):
  print('%*d => %.3f, %*s' % (3, k, mejores_sim_food_ordenadas[k][0], 40, mejores_sim_food_ordenadas[k][1]))

Similitudes con base a la evaluación de la comida con mayores valores de correlación:
  1 => 0.792,          Restaurante la Estrella de Dima
  2 => 0.700,                            pizza clasica
  3 => 0.665,                       Restaurante Guerra
  4 => 0.543,                                  El Club
  5 => 0.369,                          Pizzeria Julios
  6 => 0.311,              Restaurant Oriental Express
  7 => 0.308,            Restaurante El Cielo Potosino
  8 => 0.307,                        Hamburguesas saul
  9 => 0.283,                          Rincon Huasteco
 10 => 0.269,  la Cochinita Pibil Restaurante Yucateco


## **Ejercicio - 6**

### **De manera análoga y usando ahora la evaluación con respecto al servicio (service_rating), encuentrarás ahora los diez restaurantes con mayor similaridad al "Restaurante Pueblo Bonito" y con respecto a la métrica de similaridad de Pearson.**

### **Deberás usar la cantidad de componentes necesarias para que la varabilidad de la factorización SVD quede explicada en un 90%.**

### **a.  Para ello define primero la matriz de utilidad, UtMx_service, donde los renglones son los nombres de los restaurantes, las columnas los usuarios y las entradas las evaluaciones con respecto al servicio recibido en el restaurante.**





In [17]:
# a. Define la matriz de utilidad cuyos renglones sean los nombres de los
#    restaurantes, las columnas los IDs de los usuarios y las entradas la
#    evaluación del servicio (service_rating). La llamaremos "UtMx_service".


# ************* Inlcuye aquí tu código:*****************************


UtMx_service = df_combinado.pivot_table(values='service_rating', index="name", columns="userID", fill_value=0)


# *********** Aquí termina la sección de agregar código *************




print('Dimensión de la matriz de Utilidad sobre la evaluación del servicio:')
print('(restaurantes, usuarios) =', (UtMx_service.shape))
UtMx_service.head()

Dimensión de la matriz de Utilidad sobre la evaluación del servicio:
(restaurantes, usuarios) = (129, 138)


userID                     U1001  U1002  U1003  U1004  U1005  U1006  U1007  \
name                                                                         
Abondance Restaurante Bar    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
Arrachela Grill              0.0    0.0    0.0    0.0    0.0    0.0    0.0   
Cabana Huasteca              0.0    0.0    2.0    0.0    0.0    0.0    0.0   
Cafe Chaires                 0.0    0.0    0.0    0.0    0.0    1.0    1.0   
Cafeteria cenidet            0.0    0.0    0.0    0.0    0.0    0.0    0.0   

userID                     U1008  U1009  U1010  ...  U1129  U1130  U1131  \
name                                            ...                        
Abondance Restaurante Bar    1.0    0.0    0.0  ...    0.0    0.0    0.0   
Arrachela Grill              0.0    0.0    0.0  ...    0.0    0.0    0.0   
Cabana Huasteca              0.0    1.0    0.0  ...    0.0    0.0    0.0   
Cafe Chaires                 0.0    0.0    0.0  ...    0.0    0.0    0.0   
Cafeteria cenidet            0.0    0.0    0.0  ...    0.0    0.0    0.0   

userID                     U1132  U1133  U1134  U1135  U1136  U1137  U1138  
name                                                                        
Abondance Restaurante Bar    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
Arrachela Grill              0.0    0.0    0.0    0.0    0.0    0.0    0.0  
Cabana Huasteca              0.0    0.0    1.0    0.0    0.0    0.0    0.0  
Cafe Chaires                 0.0    0.0    0.0    0.0    0.0    0.0    0.0  
Cafeteria cenidet            0.0    0.0    0.0    0.0    0.0    0.0    0.0  

[5 rows x 138 columns]

In [33]:
# b. Incluye el código necesario para obtener y desplegar los 10 restaurantes con
#    mayor similaridad de Pearson al "Restaurante Pueblo Bonito", y usando solo
#    el 90% de las componentes que explican la varianza de la matriz UtMx_service.
#    Desplegarlos en orden descendente, sin incliur el restaurante de referencia.

# ************* Inlcuye aquí tu código:*****************************


SVD_food = TruncatedSVD( n_components = 138 )
SVD_food.fit(UtMx_service)


for j in range(138):
  if SVD_food.explained_variance_ratio_[0:j].sum() > 0.90:
    break

SVD_food = TruncatedSVD(n_components = j)
resultant_matrix_food = SVD_food.fit_transform(UtMx_service)


corr_mat_food = np.corrcoef(resultant_matrix_food)


restaurante_de_referencia = 'Restaurante Pueblo Bonito'
nombres_rest = UtMx_service.T.columns 
idx_rest = list(nombres_rest).index(restaurante_de_referencia)
corr_rest = corr_mat_food[idx_rest]


idx = (corr_rest>0)
mejor_servicio = list()
for i in range(len(nombres_rest[idx])):
  mejor_servicio.append((corr_rest[idx][i], nombres_rest[idx][i]))


mejor_servicio_sorted = sorted(mejor_servicio, reverse=True)

for k in range(1,11):
  print('%*d => %.3f, %*s' % (3, k, mejor_servicio_sorted[k][0], 40, mejor_servicio_sorted[k][1]))

# *********** Aquí termina la sección de agregar código *************

  1 => 0.666,                            pizza clasica
  2 => 0.585,                                  El Club
  3 => 0.496,                        Hamburguesas saul
  4 => 0.469,          Restaurante la Estrella de Dima
  5 => 0.430,                       Restaurante Guerra
  6 => 0.400,                     Restaurant Orizatlan
  7 => 0.383,            Restaurante El Cielo Potosino
  8 => 0.355,  la Cochinita Pibil Restaurante Yucateco
  9 => 0.335,    La Fontana Pizza Restaurante and Cafe
 10 => 0.327,              Restaurant Oriental Express


Resultados Comida

*  1 => 0.792,          Restaurante la Estrella de Dima
*  2 => 0.700,                            pizza clasica
*  3 => 0.665,                       Restaurante Guerra
*  4 => 0.543,                                  El Club
*  5 => 0.369,                          Pizzeria Julios
*  6 => 0.311,              Restaurant Oriental Express
*  7 => 0.308,            Restaurante El Cielo Potosino
*  8 => 0.307,                        Hamburguesas saul
*  9 => 0.283,                          Rincon Huasteco
* 10 => 0.269,  la Cochinita Pibil Restaurante Yucateco



Resultados Servicio

*  1 => 0.666,                            pizza clasica
*  2 => 0.585,                                  El Club
*  3 => 0.496,                        Hamburguesas saul
*  4 => 0.469,          Restaurante la Estrella de Dima
*  5 => 0.430,                       Restaurante Guerra
*  6 => 0.400,                     Restaurant Orizatlan *
*  7 => 0.383,            Restaurante El Cielo Potosino
*  8 => 0.355,  la Cochinita Pibil Restaurante Yucateco
*  9 => 0.335,    La Fontana Pizza Restaurante and Cafe *
* 10 => 0.327,              Restaurant Oriental Express

## **Ejercicio - 7**

### **Incluye tus comentarios y conclusiones de la actividad. En particular indica cuáles restaurantes encontraste con correlaciones altas tanto para la calificación de la comida, como del servicio.**

++++++++ Inicia la sección de agregar texto: +++++++++++

En general hay una similitud entre ambas, tanto para la matrix de puntajes de comida con la de servicio salvo dos restaurantes que aparecieron en los resultados del segundo experimento (Restaurant Orizatlan y La Fontana Pizza Restaurante and Cafe).

++++++++ Termina la sección de agregar texto. +++++++++++

>> **Fin de la Actividad de Sistemas de Recomendación**